pkgs/cronet_http/lib/src/cronet_client.dart

```diff
@@ -186,26 +212,114 @@ class CronetClient extends BaseClient {
           'HTTP request failed. Client is already closed.', request.url);
     }
 
-    if (_engine == null) {
-      // Create the future here rather than in the [fromCronetEngineFuture]
-      // factory so that [close] does not have to await the future just to
-      // close it in the case where [send] is never called.
-      //
-      // Assign to _engineFuture instead of just `await`ing the result of
-      // `CronetEngine.build()` to prevent concurrent executions of `send`
-      // from creating multiple [CronetEngine]s.
-      _engineFuture ??= CronetEngine.build();
-      try {
-        _engine = await _engineFuture;
-      } catch (e) {
-        throw ClientException(
-            'Exception building CronetEngine: ${e.toString()}', request.url);
-      }
-    }
+    _engine ??= CronetEngine.build();
 
     final stream = request.finalize();
-
     final body = await stream.toBytes();
+    final responseCompleter = Completer<StreamedResponse>();
+    final engine = _engine!._engine;
+
+    late jb.UrlRequest cronetRequest;
+    var numRedirects = 0;
+    StreamController<List<int>>? responseStream;
+    jb.ByteBuffer? byteBuffer;
+
+    // The order of callbacks generated by Cronet is documented here:
+    // https://developer.android.com/guide/topics/connectivity/cronet/lifecycle
+
+    final cronetCallbacks =
+        jb.UrlRequestCallbackProxy_UrlRequestCallbackInterface.implement(
+            jb.$UrlRequestCallbackProxy_UrlRequestCallbackInterfaceImpl(
+      onResponseStarted: (urlRequest, responseInfo) {
+        responseStream = StreamController();
+        final responseHeaders =
+            _cronetToClientHeaders(responseInfo.getAllHeaders());
+        int? contentLength;
+
+        switch (responseHeaders['content-length']) {
+          case final contentLengthHeader?
+              when !_digitRegex.hasMatch(contentLengthHeader):
+            responseCompleter.completeError(ClientException(
+              'Invalid content-length header [$contentLengthHeader].',
+              request.url,
+            ));
+            urlRequest.cancel();
+            return;
+          case final contentLengthHeader?:
+            contentLength = int.parse(contentLengthHeader);
+        }
+        responseCompleter.complete(StreamedResponse(
+          responseStream!.stream,
+          responseInfo.getHttpStatusCode(),
+          contentLength: contentLength,
+          reasonPhrase: responseInfo
+              .getHttpStatusText()
+              .toDartString(releaseOriginal: true),
+          request: request,
+          isRedirect: false,
+          headers: responseHeaders,
+        ));
+
+        byteBuffer = jb.ByteBuffer.allocateDirect(1024 * 1024);
```

Comment:
[out of scope for this PR] Consider adding a constant for this number - it looks like it is repeated a few times

Reply:
ACK (I think that this disappears in my next PR)